libs/src/services/discoveryProvider/index.js

```diff
@@ -473,30 +472,94 @@ class DiscoveryProvider {
   // endpoint - base route
   // urlParams - string of url params to be appended after base route
   // queryParams - object of query params to be appended to url
-  async _makeRequest (requestObj, retries = MAKE_REQUEST_RETRY_COUNT, attempedRetries = 0) {
-    if (attempedRetries > MAX_MAKE_REQUEST_RETRY_COUNT) {
-      console.error('Attempted max request retries.')
+  async _makeRequest (requestObj, attemptedRetries = 0) {
+    try {
+      const newDiscProvEndpoint = await this.getHealthyDiscoveryProviderEndpoint(attemptedRetries)
+
+      // If new DP endpoint is selected, update disc prov endpoint and reset attemptedRetries count
+      if (this.discoveryProviderEndpoint !== newDiscProvEndpoint) {
+        console.info(`Current Discovery Provider endpoint ${this.discoveryProviderEndpoint} is unhealthy. Switching over to
+          the new Discovery Provider endpoint ${newDiscProvEndpoint}!`)
+        this.discoveryProviderEndpoint = newDiscProvEndpoint
+        attemptedRetries = 0
+      }
+    } catch (e) {
+      console.error(e.message)
       return
     }
 
-    if (!this.discoveryProviderEndpoint) {
-      const endpoint = await this.serviceSelector.select()
-      this.setEndpoint(endpoint)
+    let axiosRequest = this.createDiscProvRequest(requestObj)
+
+    let response
+    let parsedResponse
+    try {
+      response = await axios(axiosRequest)
+      parsedResponse = Utils.parseDataFromResponse(response)
+    } catch (e) {
+      console.error(`Failed to make Discovery Provider request: ${e}`)
+      return this._makeRequest(requestObj, attemptedRetries + 1)
     }
 
-    if (retries === 0) {
-      // Reset the retries count in the case that the newly selected disc prov fails, we can
-      // allow it to try MAKE_REQUEST_RETRIES_COUNT number of times before trying another
-      retries = MAKE_REQUEST_RETRY_COUNT
-      const endpoint = await this.serviceSelector.select()
-      this.setEndpoint(endpoint)
+    if (
+      this.ethContracts &&
+      !this.ethContracts.isInRegressedMode() &&
+      'latest_indexed_block' in parsedResponse &&
+      'latest_chain_block' in parsedResponse
+    ) {
+      const {
+        latest_indexed_block: indexedBlock,
+        latest_chain_block: chainBlock
+      } = parsedResponse
+
+      if (
+        !chainBlock ||
+        !indexedBlock ||
+        (chainBlock - indexedBlock) > UNHEALTHY_BLOCK_DIFF
+      ) {
+        // If disc prov is an unhealthy num blocks behind,
+        console.info(`${this.discoveryProviderEndpoint} is too far behind. Retrying request...`)
+        this._makeRequest(requestObj, attemptedRetries + 1)
```

Comment:
should this return this._makeRequest? i dont think we want to return the parsedResponse in this case

Reply:
ah good catch